src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -0,0 +1,369 @@
+
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        throw;
+                    }
+
+                    isCanceled = true;
+                }
+
+                return new ReadResult(GetCurrentReadOnlySequence(), isCanceled, _isStreamCompleted);
+            }
+        }
+
+        private void ClearCancellationToken()
+        {
+            lock (_lock)
+            {
+                _internalTokenSource = null;
+            }
+        }
+
+        private void ThrowIfCompleted()
+        {
+            if (_isReaderCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+            }
+        }
+
+        public override bool TryRead(out ReadResult result)
+        {
+            ThrowIfCompleted();
+
+            return TryReadInternal(InternalTokenSource, out result);
+        }
+
+        private bool TryReadInternal(CancellationTokenSource source, out ReadResult result)
+        {
+            bool isCancellationRequested = source.IsCancellationRequested;
+            if (isCancellationRequested || _bufferedBytes > 0 && (!_examinedEverything || _isStreamCompleted))
+            {
+                // If TryRead/ReadAsync are called and cancellation is requested, we need to make sure memory is allocated for the ReadResult,
+                // otherwise if someone calls advance afterward on the ReadResult, it will throw.
+                if (isCancellationRequested)
+                {
+                    AllocateReadTail();
+
+                    ClearCancellationToken();
+                }
+
+                result = new ReadResult(GetCurrentReadOnlySequence(), isCancellationRequested, _isStreamCompleted);
+                return true;
+            }
+
+            result = new ReadResult();
+            return false;
+        }
+
+        private ReadOnlySequence<byte> GetCurrentReadOnlySequence()
+        {
+            return new ReadOnlySequence<byte>(_readHead, _readIndex, _readTail, _readTail.End);
+        }
+
+        private void AllocateReadTail()
+        {
+            if (_readHead == null)
+            {
+                Debug.Assert(_readTail == null);
+                _readHead = AllocateSegment();
+                _readTail = _readHead;
+            }
+            else if (_readTail.WritableBytes < _minimumReadThreshold)
+            {
+                CreateNewTailSegment();
+            }
+        }
+
+        private void CreateNewTailSegment()
+        {
+            BufferSegment nextSegment = AllocateSegment();
+            _readTail.SetNext(nextSegment);
+            _readTail = nextSegment;
+        }
+
+        private BufferSegment AllocateSegment()
+        {
+            BufferSegment nextSegment = CreateSegmentUnsynchronized();
+
+            if (_pool is null)
+            {
+                nextSegment.SetOwnedMemory(ArrayPool<byte>.Shared.Rent(_bufferSize));
+            }
+            else
```

Comment:
Do we need a check for pools MaxSegmentSize here like we do in the Pipe.cs? Should this code be shared too?

Reply:
https://github.com/dotnet/corefx/pull/36575/files#diff-c30bc95b551fe06801a5132f131b11ecR54

Reply:
Interesting, did you consider throwing in that case?

Reply:
It could but I'm not sure it should, it would mean to avoid it the caller would do the math instead of us.

Reply:
At least callers would know that their expectations are incorrect.

Reply:
I'm not convinced yet. We could do the same with each of the properties. MinimumReadSize needs to be <= BufferSize and BufferSize should be <= MaxBufferSize